### PR TITLE
Task/add modal

### PR DESCRIPTION
**Pull request by jduong09**
Added a modal that would prompt the user to leave the current form they are on, and discard the changes they have been working on.

### CODE REVIEW

**Comment by jduong09** [comment]
### Features
- Added modal to handle when user is switching from a form (create/update) to a different view.
- Prompts user to stay on current view, or discard changes and go to the selected view.

### Still needs addressing
- Location of modal
- Styling of modal

### Additional Feature
- Add modal for deleting subscriptions. Prompt user to make sure they don't accidentally delete a subscription.

### Tests Done
- View: Dashboard Calendar
- [ ] Displays Subscription List, shows dashboardCalendar
- View: Subscription Detail
- [ ] Clicking Edit button shows Update Subscription form (no modal)
- [ ] Clicking Close shows Dashboard Calendar (no modal) 
- [ ] Clicking Delete shows Dashboard Calendar (no modal)
- [ ] Clicking Edit Pen shows Update Subscription Form (no modal).
- View: Update Subscription Form
- [ ] Clicking close button shows modal
    - [ ] Clicking stay, returns to form
    - [ ] Clicking discard changes, directs to Subscription Detail
- [ ] Clicking Edit Pen
    - [ ] Shows modal
    - [ ] Stays in form if user click's 'Stay In Page'
    - [ ] Directs to correct form if user click's 'Discard Changes'
- [ ] Clicking Delete in subscription list deletes correct subscription
    - [ ] Stays on Update Subscription Form is deleting different subscription
    - [ ] Redirects to Dashboard Calendar if deleting the subscription that is being updated.
- View : Create Subscription Form
    - [ ] Close Button Shows Modal
    - [ ] Same tests as other Modal tests.
    - [ ] Clicking subscription from subscription list should show modal.
    - [ ] clicking Edit Pen from subscription list should show modal.
    - [ ] Clicking Delete icon should keep create subscription form as is.

File: client/src/components/dashboardModal.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import React  from 'react';
+
+const DashboardModal = () => (
+  <div className="modal fade" id="myModal" tabIndex="-1" aria-labelledby="myModal" aria-hidden="true">
```
Comment (ziggysauce) [inline client/src/components/dashboardModal.js]: using a passed in ID would allow for this modal to be more usable across the app. But since it's only used for this one functionality right now I think this is fine. Though I would recommend using more meaningful IDs in this file.

File: client/src/components/dashboardModal.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <div className="modal-dialog">
+      <div className="modal-content">
+        <div className="modal-header">
+          <h5 className="modal-title" id="exampleModalLabel">Navigating</h5>
+        </div>
+        <div className="modal-body">
+          Are you sure you want to navigate away from your current page?
```
Comment (ziggysauce) [inline client/src/components/dashboardModal.js]: can do some re-wording here

```suggestion
          <h5 className="modal-title" id="exampleModalLabel">Are you sure?</h5>
        </div>
        <div className="modal-body">
          If you leave all unsaved changes will be discarded. 
```

File: client/src/components/user/dashboard.js [hunk 3]
```diff
@@ -9,12 +10,27 @@ import DashboardCalendar from '../date/dashboardCalendar';
...
 import SubscriptionDetail from '../subscription/subscriptionDetail';
+import DashboardModal from '../dashboardModal';
 import logo from '../../assets/watering-can.png';
 
 const href = process && process.env && process.env.NODE_ENV === 'production'
   ? '/auth/logout'
   : 'http://localhost:5000/auth/logout';
 
+const waitUserClick = () => (
+  new Promise((resolve, ) => {
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js]: this looks like it only needs one argument

```suggestion
  new Promise((resolve) => {
```

File: client/src/components/user/dashboard.js [hunk 4]
```diff
@@ -9,12 +10,27 @@ import DashboardCalendar from '../date/dashboardCalendar';
...
 import CreateSubscription from '../subscription/createSubscription';
 import SubscriptionDetail from '../subscription/subscriptionDetail';
+import DashboardModal from '../dashboardModal';
 import logo from '../../assets/watering-can.png';
 
 const href = process && process.env && process.env.NODE_ENV === 'production'
   ? '/auth/logout'
   : 'http://localhost:5000/auth/logout';
 
+const waitUserClick = () => (
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js]: is all of this necessary here? couldn't you just pass up the event click from the modal to whichever parent component is using it? In your case it looks like it would be handled in `handleDashboardChange`

File: client/src/components/user/dashboard.js [hunk 5]
```diff
@@ -72,10 +89,29 @@ class Dashboard extends React.Component {
...
+
+    const modal = new Modal(document.getElementById('myModal'));
+    modal.show();
+
+    const userInput = await waitUserClick();
+    if (userInput === 'next') {
+      await this.setState({ mainComponentView: newView });
+    } else if (userInput === 'close') {
+      await this.setState({ mainComponentView });
+    }
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js]: I don't think you need to `await` setState since that function doesn't handle promises

File: client/src/components/user/dashboard.js [hunk 6]
```diff
@@ -72,10 +89,29 @@ class Dashboard extends React.Component {
...
+    return modal.hide();
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js]: I don't think you need to `return` here since this is just the end of the function

```suggestion
    modal.hide();
```

File: client/src/components/user/dashboard.js [hunk 7]
```diff
@@ -72,10 +89,29 @@ class Dashboard extends React.Component {
...
   }
 
+  setMainComponentView = async (newView) => {
+    await this.setState({ mainComponentView: newView });
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js]: same as above here and everywhere else that you're doing it, i don't think you need the async await when using the `this.setState()` cal

File: client/src/components/user/dashboard.js [hunk 8]
```diff
@@ -139,40 +178,77 @@ class Dashboard extends React.Component {
...
+            <div className="d-md-none">
+              <DashboardCalendar subscriptions={subscriptions} />
+            </div>
           </div>
-          <div className="d-md-none">
-            <DashboardCalendar subscriptions={subscriptions}/>
+        );
+      case 'updateSubscription': 
+        return (
+          <div>
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js:223]: is this outer div necessary? seems like we can get rid of it

File: client/src/components/user/dashboard.js [hunk 9]
```diff
@@ -139,40 +178,77 @@ class Dashboard extends React.Component {
...
+            <div className="p-3 m-2 d-flex flex-wrap borderSubscriptionForm d-none d-md-block">
+              <div className="col d-flex justify-content-between align-items-center">
+                <div />
```
Comment (ziggysauce) [inline client/src/components/user/dashboard.js:226]: what's going on with this div?

**Review by ziggysauce — CHANGES_REQUESTED** [review]
excellent work here bro! just some minor changes and this should be good to merge in

File: client/src/components/dashboardModal.js (hunk 1 above)
Comment (jduong09) [reply client/src/components/dashboardModal.js]: Gotcha. That makes sense. And I was looking towards making a modal for Deleting too. So meaningful ID would make a lot more sense.

File: client/src/components/user/dashboard.js (hunk 8 above)
Comment (jduong09) [reply client/src/components/user/dashboard.js:223]: The outer div holds the two separate divs. One div displays on medium screens and bigger, while the other displays on screens smaller than medium. Without the div, they are separate divs, and you cannot return two sibling html elements.

File: client/src/components/user/dashboard.js (hunk 9 above)
Comment (jduong09) [reply client/src/components/user/dashboard.js:226]: the parent div is a Flexbox container, and the empty div places empty space in the first spot, resulting in the title in the middle and the close button at the end of the row.

File: client/src/components/user/dashboard.js (hunk 8 above)
Comment (ziggysauce) [reply client/src/components/user/dashboard.js:223]: oh i see, interesting. yeah that wasn't clear to me before since there weren't any classes on this outer div, but what you're saying makes sense with the other code.

File: client/src/components/user/dashboard.js (hunk 6 above)
Comment (jduong09) [reply client/src/components/user/dashboard.js]: return modal.hide is used because i have an if-statement that returns from the function early. ESLINT throws an error if you do not have a return statement at the end due to this.

**Comment by jduong09** [comment]
### Changes
- Changed dashboardModal component to modalComponent, for reusability. If we need more modals, we can utilize modalComponent.

File: client/src/components/user/dashboard.js (hunk 6 above)
Comment (jduong09) [reply client/src/components/user/dashboard.js]: https://eslint.org/docs/rules/consistent-return

**Comment by ziggysauce** [comment]
> ### Changes
> * Changed dashboardModal component to modalComponent, for reusability. If we need more modals, we can utilize modalComponent.

this is good in practice, but the content is still static to just navigation right now. if we. could dynamically render the header, body, and footer content then this reusability would be more applicable. But until then, keeping this isolated to a modal that specifically is used for confirming the navigation for the user is fine and can be named applicable to that functionality

**Review by ziggysauce — CHANGES_REQUESTED** [review]
I'm seeing this when trying to delete a subscription:
Screen Shot 2022-03-24 at 8 29 35 PM

File: client/src/components/modalComponent.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        </div>
+        <div className="modal-body">
+          {id === 'deleteModal'
+            ? 'Are you sure you want to delete this subscription?'
+            : 'If you leave, all unsaved changes will be discarded.'
+          }
+        </div>
+        <div className="modal-footer">
+          <button onClick={() => handleModalClick(false)} type="button" className="btn btn-primary" data-bs-dismiss="modal">
+            {id === 'deleteModal' ? 'Do not Delete' : 'Stay On Page'}
```
Comment (ziggysauce) [inline client/src/components/modalComponent.js]: hahahah this is too wordy. we already have "Delete" in the main question above as well as in the other button

```suggestion
            {id === 'deleteModal' ? 'Cancel' : 'Stay On Page'}
```
